File: packages/google-pubsub/src/consumer.ts
/**
 * Google Cloud Pub/Sub Consumer Implementation
 */

import {
  PubSub,
  type Subscription,
  type Message as PubSubMessage,
} from '@google-cloud/pubsub';
import {
  BaseConsumer,
  createMessage,
  type MessageHandler,
  type BatchMessageHandler,
  type SubscribeOptions,
  type HealthStatus,
  ConnectionError,
} from '@anyq/core';
import type { PubSubConsumerConfig } from './config.js';
import { PUBSUB_DEFAULTS } from './config.js';

/**
 * Google Cloud Pub/Sub Consumer
 *
 * Consumes messages from Google Cloud Pub/Sub subscriptions
 */
export class PubSubConsumer<T = unknown> extends BaseConsumer<T> {
  private client: PubSub | null = null;
  private subscription: Subscription | null = null;
  private readonly pubsubConfig: PubSubConsumerConfig;

  constructor(config: PubSubConsumerConfig) {
    super(config);
    this.pubsubConfig = {
      ...config,
      subscription: {
        ...PUBSUB_DEFAULTS.subscription,
        ...config.subscription,
      },
    };
  }

  /**
   * Connect to Google Cloud Pub/Sub
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      const { connection, subscription: subConfig } = this.pubsubConfig;

      // Configure client options
      const clientOptions: {
        projectId: string;
        keyFilename?: string;
        apiEndpoint?: string;
      } = {
        projectId: connection.projectId,
      };

      if (connection.keyFilename) {
        clientOptions.keyFilename = connection.keyFilename;
      }

      if (connection.apiEndpoint) {
        clientOptions.apiEndpoint = connection.apiEndpoint;
      }

      // Set emulator environment variable if needed
      if (connection.emulatorMode && connection.apiEndpoint) {
        process.env.PUBSUB_EMULATOR_HOST = connection.apiEndpoint.replace('http://', '');
      }

      this.client = new PubSub(clientOptions);

      // Get subscription
      this.subscription = this.client.subscription(subConfig.name, {
        flowControl: {
          maxMessages: subConfig.flowControl?.maxMessages ?? PUBSUB_DEFAULTS.flowControl.maxMessages,
          allowExcessMessages: false,
        },
        ackDeadline: subConfig.ackDeadlineSeconds ?? PUBSUB_DEFAULTS.subscription.ackDeadlineSeconds,
      });

      if (subConfig.autoCreate) {
        await this.ensureSubscription();
      }

      this._connected = true;
      this.logger.info('Pub/Sub consumer connected', {
        projectId: connection.projectId,
        subscription: subConfig.name,
      });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to Pub/Sub',
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Ensure the subscription exists
   */
  private async ensureSubscription(): Promise<void> {
    if (!this.subscription || !this.client) {
      throw new Error('Subscription not initialized');
    }

    const { subscription: subConfig, topicName } = this.pubsubConfig;

    try {
      const [exists] = await this.subscription.exists();
      if (!exists) {
        if (!topicName) {
          throw new Error('Topic name required for auto-creating subscription');
        }

        // Ensure topic exists first
        const topic = this.client.topic(topicName);
        const [topicExists] = await topic.exists();
        if (!topicExists) {
          await topic.create();
          this.logger.info('Topic created', { topic: topicName });
        }

        // Create subscription
        const [createdSub] = await topic.createSubscription(subConfig.name, {
          ackDeadlineSeconds: subConfig.ackDeadlineSeconds,
          deadLetterPolicy: subConfig.deadLetterPolicy ? {
            deadLetterTopic: `projects/${this.pubsubConfig.connection.projectId}/topics/${subConfig.deadLetterPolicy.deadLetterTopic}`,
            maxDeliveryAttempts: subConfig.deadLetterPolicy.maxDeliveryAttempts,
          } : undefined,
          retryPolicy: subConfig.retryPolicy ? {
            minimumBackoff: { seconds: subConfig.retryPolicy.minimumBackoff },
            maximumBackoff: { seconds: subConfig.retryPolicy.maximumBackoff },
          } : undefined,
        });

        this.subscription = createdSub;
        this.logger.info('Subscription created', { subscription: subConfig.name });
      }
    } catch (error) {
      // Ignore if subscription already exists (race condition)
      const err = error as { code?: number };
      if (err.code !== 6) { // 6 = ALREADY_EXISTS
        throw error;
      }
    }
  }

  /**
   * Disconnect from Pub/Sub
   */
  async disconnect(): Promise<void> {
    if (this.subscription) {
      await this.subscription.close();
      this.subscription = null;
    }

    if (this.client) {
      await this.client.close();
      this.client = null;
    }

    this._connected = false;
    this.logger.info('Pub/Sub consumer disconnected');
  }

  /**
   * Subscribe to messages
   */
  async subscribe(
    handler: MessageHandler<T>,
    _options?: SubscribeOptions
  ): Promise<void> {
    if (!this.subscription) {
      throw new ConnectionError('Consumer not connected');
    }

    this.subscription.on('message', async (message: PubSubMessage) => {
      if (this._paused) {
        // Nack to re-deliver later when paused
        message.nack();
        return;
      }

      try {
        const convertedMessage = this.convertMessage(message);
        await handler(convertedMessage);
      } catch (error) {
        this.logger.error('Error processing message', { error });
        // Don't ack - message will be redelivered
      }
    });

    this.subscription.on('error', (error: Error) => {
      this.logger.error('Subscription error', { error: error.message });
      this.emit('error', error);
    });

    this.logger.info('Subscribed to messages', {
      subscription: this.pubsubConfig.subscription.name,
    });
  }

  /**
   * Subscribe to messages in batches
   */
  async subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.subscription) {
      throw new ConnectionError('Consumer not connected');
    }

    const batchSize = options?.batchSize ?? 10;
    const batchTimeout = options?.batchTimeout ?? 5000;
    let batch: PubSubMessage[] = [];
    let batchTimer: ReturnType<typeof setTimeout> | null = null;

    const processBatch = async () => {
      if (batch.length === 0) return;

      const currentBatch = [...batch];
      batch = [];

      if (batchTimer) {
        clearTimeout(batchTimer);
        batchTimer = null;
      }

      try {
        const messages = currentBatch.map(msg => this.convertMessage(msg));
        await handler(messages);
      } catch (error) {
        this.logger.error('Error processing batch', { error });
        // Nack all messages in the batch
        for (const msg of currentBatch) {
          msg.nack();
        }
      }
    };

    this.subscription.on('message', async (message: PubSubMessage) => {
      if (this._paused) {
        message.nack();
        return;
      }

      batch.push(message);

      if (batch.length >= batchSize) {
        await processBatch();
      } else if (!batchTimer) {
        batchTimer = setTimeout(() => processBatch(), batchTimeout);
      }
    });

    this.subscription.on('error', (error: Error) => {
      this.logger.error('Subscription error', { error: error.message });
      this.emit('error', error);
    });

    this.logger.info('Subscribed to messages (batch mode)', {
      subscription: this.pubsubConfig.subscription.name,
      batchSize,
      batchTimeout,
    });
  }

  /**
   * Convert Pub/Sub message to IMessage format
   */
  private convertMessage(message: PubSubMessage) {
    const data = message.data.toString('utf8');
    let body: T;

    try {
      body = JSON.parse(data) as T;
    } catch {
      body = data as unknown as T;
    }

    // Extract headers from attributes
    const headers: Record<string, string | Buffer | undefined> = {};
    if (message.attributes) {
      for (const [key, value] of Object.entries(message.attributes)) {
        headers[key] = value;
      }
    }

    const messageId = message.id;
    const publishTime = message.publishTime;

    return createMessage<T>({
      id: messageId,
      body,
      headers,
      timestamp: publishTime,
      deliveryAttempt: message.deliveryAttempt ?? 1,
      metadata: {
        provider: 'google-pubsub',
        googlePubsub: {
          subscription: this.pubsubConfig.subscription.name,
          ackId: message.ackId,
          publishTime,
          orderingKey: message.orderingKey,
        },
      },
      raw: message,
      onAck: async () => {
        message.ack();
        this.logger.debug('Message acknowledged', { messageId });
      },
      onNack: async () => {
        message.nack();
        this.logger.debug('Message negatively acknowledged', { messageId });
      },
      onExtendDeadline: async (_seconds: number) => {
        // Pub/Sub modifyAckDeadline requires subscription-level operation
        // For simplicity, we just log the extension request
        this.logger.debug('Message deadline extension requested', { messageId });
      },
    });
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();

    try {
      if (!this._connected || !this.subscription) {
        return {
          healthy: false,
          connected: false,
          error: 'Not connected',
        };
      }

      // Check if subscription exists
      const [exists] = await this.subscription.exists();

      return {
        healthy: exists,
        connected: this._connected,
        latencyMs: Date.now() - start,
        details: {
          subscription: this.pubsubConfig.subscription.name,
          projectId: this.pubsubConfig.connection.projectId,
          paused: this._paused,
        },
      };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Create a Pub/Sub consumer instance
 */
export function createPubSubConsumer<T = unknown>(
  config: PubSubConsumerConfig
): PubSubConsumer<T> {
  return new PubSubConsumer<T>(config);
}

File: go/pubsub/consumer.go
package pubsub

import (
	"context"
	"fmt"
	"time"

	"cloud.google.com/go/pubsub"

	"github.com/sns45/anyq/go/core"
)

// Consumer receives messages from a Google Cloud Pub/Sub subscription.
//
// Pub/Sub uses streaming pull: Subscription.Receive invokes the callback
// concurrently from multiple goroutines. Each delivery is wrapped in its own
// core.Message (with per-message OnAck/OnNack), so there is no shared mutable
// per-message state across callbacks. ApplyStrategy operates only on the message
// passed to it, so concurrent invocation is safe; the only shared state
// (pause flag, circuit breaker, logger) is already synchronised by the core
// base types.
type Consumer struct {
	*core.BaseConsumer
	cfg    ConsumerConfig
	client *pubsub.Client
	sub    *pubsub.Subscription
}

// NewConsumer builds a Pub/Sub consumer.
func NewConsumer(cfg ConsumerConfig) *Consumer {
	if cfg.Driver == "" {
		cfg.Driver = core.DriverGooglePubSub
	}
	bc := &core.BaseConsumer{}
	bc.InitBase(cfg.BaseQueueConfig)

	c := &Consumer{BaseConsumer: bc, cfg: cfg}
	bc.Bind(c)
	return c
}

// Connect creates the client and ensures the subscription exists.
func (c *Consumer) Connect(ctx context.Context) error {
	if c.IsConnected() {
		return nil
	}
	if err := c.VerifyParkPolicy(); err != nil {
		return err
	}

	client, err := pubsub.NewClient(ctx, c.cfg.Connection.ProjectID, clientOptions(c.cfg.Connection)...)
	if err != nil {
		return core.NewConnectionError("failed to connect to Pub/Sub", err)
	}
	c.client = client

	sub := client.Subscription(c.cfg.Subscription.Name)
	fc := c.cfg.Subscription.FlowControl
	if fc != nil && fc.MaxMessages > 0 {
		sub.ReceiveSettings.MaxOutstandingMessages = fc.MaxMessages
	} else {
		sub.ReceiveSettings.MaxOutstandingMessages = defaultFlowControlMessages
	}
	if fc != nil && fc.MaxBytes > 0 {
		sub.ReceiveSettings.MaxOutstandingBytes = fc.MaxBytes
	} else {
		sub.ReceiveSettings.MaxOutstandingBytes = defaultFlowControlBytes
	}
	c.sub = sub

	if boolOr(c.cfg.Subscription.AutoCreate, true) {
		if err := c.ensureSubscription(ctx); err != nil {
			_ = client.Close()
			c.client = nil
			c.sub = nil
			return core.NewConnectionError("failed to ensure subscription", err)
		}
	}

	c.SetConnected(true)
	c.Logger.Info("Pub/Sub consumer connected", map[string]any{
		"projectId":    c.cfg.Connection.ProjectID,
		"subscription": c.cfg.Subscription.Name,
	})
	return nil
}

func (c *Consumer) ensureSubscription(ctx context.Context) error {
	exists, err := c.sub.Exists(ctx)
	if err != nil {
		return err
	}
	if exists {
		return nil
	}

	if c.cfg.TopicName == "" {
		return fmt.Errorf("topic name required for auto-creating subscription")
	}

	// Ensure the topic exists first.
	topic := c.client.Topic(c.cfg.TopicName)
	topicExists, err := topic.Exists(ctx)
	if err != nil {
		return err
	}
	if !topicExists {
		topic, err = c.client.CreateTopic(ctx, c.cfg.TopicName)
		if err != nil && !isAlreadyExists(err) {
			return err
		}
		if err == nil {
			c.Logger.Info("Topic created", map[string]any{"topic": c.cfg.TopicName})
		}
	}

	subCfg := pubsub.SubscriptionConfig{Topic: topic}
	if d := c.cfg.Subscription.AckDeadlineSeconds; d > 0 {
		subCfg.AckDeadline = time.Duration(d) * time.Second
	} else {
		subCfg.AckDeadline = time.Duration(defaultAckDeadlineSeconds) * time.Second
	}
	subCfg.EnableExactlyOnceDelivery = c.cfg.Subscription.ExactlyOnceDelivery
	if dlp := c.cfg.Subscription.DeadLetterPolicy; dlp != nil && dlp.DeadLetterTopic != "" {
		subCfg.DeadLetterPolicy = &pubsub.DeadLetterPolicy{
			DeadLetterTopic:     fmt.Sprintf("projects/%s/topics/%s", c.cfg.Connection.ProjectID, dlp.DeadLetterTopic),
			MaxDeliveryAttempts: dlp.MaxDeliveryAttempts,
		}
	}
	if rp := c.cfg.Subscription.RetryPolicy; rp != nil {
		subCfg.RetryPolicy = &pubsub.RetryPolicy{
			MinimumBackoff: time.Duration(rp.MinimumBackoffSeconds) * time.Second,
			MaximumBackoff: time.Duration(rp.MaximumBackoffSeconds) * time.Second,
		}
	}

	created, err := c.client.CreateSubscription(ctx, c.cfg.Subscription.Name, subCfg)
	if err != nil {
		if isAlreadyExists(err) {
			return nil
		}
		return err
	}
	created.ReceiveSettings = c.sub.ReceiveSettings
	c.sub = created
	c.Logger.Info("Subscription created", map[string]any{"subscription": c.cfg.Subscription.Name})
	return nil
}

// Disconnect closes the client. The active Receive call is stopped by cancelling
// the context passed to Subscribe.
func (c *Consumer) Disconnect(ctx context.Context) error {
	if !c.IsConnected() {
		return nil
	}
	if c.client != nil {
		_ = c.client.Close()
		c.client = nil
	}
	c.sub = nil
	c.SetConnected(false)
	c.Logger.Info("Pub/Sub consumer disconnected", nil)
	return nil
}

// Subscribe consumes messages until ctx is cancelled. The callback is invoked
// concurrently by the Pub/Sub client.
func (c *Consumer) Subscribe(ctx context.Context, handler core.Handler, opts *core.SubscribeOptions) error {
	if c.sub == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	c.Logger.Info("Subscribed to messages", map[string]any{"subscription": c.cfg.Subscription.Name})

	err := c.sub.Receive(ctx, func(cbCtx context.Context, m *pubsub.Message) {
		if c.IsPaused() {
			// Nack to redeliver later while paused.
			m.Nack()
			return
		}
		msg := c.wrap(m)
		c.handleOne(cbCtx, handler, msg, m)
	})
	if err != nil && ctx.Err() == nil {
		c.Logger.Error("Subscription error", map[string]any{"error": err.Error()})
		c.emit(err)
	}
	return err
}

func (c *Consumer) handleOne(ctx context.Context, handler core.Handler, msg core.Message, raw *pubsub.Message) {
	if herr := handler(ctx, msg); herr != nil {
		reinvoke := func() error { return handler(ctx, msg) }
		handled, applyErr := c.ApplyStrategy(ctx, msg, herr, reinvoke)
		if applyErr != nil {
			// Fail decision or context cancellation: log; ApplyStrategy has not
			// acked/nacked, so the message redelivers per the subscription policy.
			c.Logger.Error("retry strategy terminated processing", map[string]any{
				"messageId": msg.ID(), "error": applyErr.Error(),
			})
			raw.Nack()
			return
		}
		if handled {
			return
		}
		// Legacy behaviour: nack so the message is redelivered.
		c.Logger.Error("Error processing message", map[string]any{"messageId": msg.ID(), "error": herr.Error()})
		if c.Config.OnError != nil {
			c.Config.OnError(herr)
		}
		raw.Nack()
		return
	}
	// Success: ack.
	_ = msg.Ack(ctx)
}

// SubscribeBatch buffers messages into batches and dispatches them to handler.
// Pub/Sub delivers messages one at a time via Receive, so batches are assembled
// in-process.
func (c *Consumer) SubscribeBatch(ctx context.Context, handler core.BatchHandler, opts *core.SubscribeOptions) error {
	if c.sub == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}
	batchSize := 10
	if opts != nil && opts.BatchSize > 0 {
		batchSize = opts.BatchSize
	}
	batchTimeout := 5 * time.Second
	if opts != nil && opts.BatchTimeout > 0 {
		batchTimeout = opts.BatchTimeout
	}

	b := newBatcher(c, handler, batchSize, batchTimeout)
	defer b.stop()

	c.Logger.Info("Subscribed to messages (batch mode)", map[string]any{
		"subscription": c.cfg.Subscription.Name,
		"batchSize":    batchSize,
		"batchTimeout": batchTimeout.String(),
	})

	err := c.sub.Receive(ctx, func(cbCtx context.Context, m *pubsub.Message) {
		if c.IsPaused() {
			m.Nack()
			return
		}
		b.add(cbCtx, c.wrap(m), m)
	})
	if err != nil && ctx.Err() == nil {
		c.Logger.Error("Subscription error", map[string]any{"error": err.Error()})
		c.emit(err)
	}
	return err
}

// wrap converts a Pub/Sub message into a core.Message with ack/nack wired back.
func (c *Consumer) wrap(m *pubsub.Message) core.Message {
	headers := core.MessageHeaders{}
	for k, v := range m.Attributes {
		headers[k] = []byte(v)
	}

	deliveryAttempt := 1
	if m.DeliveryAttempt != nil && *m.DeliveryAttempt > 0 {
		deliveryAttempt = *m.DeliveryAttempt
	}

	return core.NewMessage(core.MessageParams{
		ID:              m.ID,
		Body:            m.Data,
		Key:             m.OrderingKey,
		Headers:         headers,
		Timestamp:       m.PublishTime,
		DeliveryAttempt: deliveryAttempt,
		Metadata: core.ProviderMetadata{
			Provider: core.DriverGooglePubSub,
			GooglePubSub: &core.GooglePubSubMetadata{
				Subscription: c.cfg.Subscription.Name,
				PublishTime:  m.PublishTime,
				OrderingKey:  m.OrderingKey,
			},
		},
		Raw: m,
		OnAck: func(context.Context) error {
			m.Ack()
			c.Logger.Debug("Message acknowledged", map[string]any{"messageId": m.ID})
			return nil
		},
		OnNack: func(_ context.Context, _ bool) error {
			m.Nack()
			c.Logger.Debug("Message negatively acknowledged", map[string]any{"messageId": m.ID})
			return nil
		},
	})
}

// emit invokes the configured OnError callback (the Go analogue of the TS
// 'error' event).
func (c *Consumer) emit(err error) {
	if c.Config.OnError != nil {
		c.Config.OnError(err)
	}
}

// HealthCheck reports consumer health by verifying the subscription exists.
func (c *Consumer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	start := time.Now()
	if !c.IsConnected() || c.sub == nil {
		return core.HealthStatus{Healthy: false, Connected: false, Error: "Not connected"}, nil
	}
	exists, err := c.sub.Exists(ctx)
	if err != nil {
		return core.HealthStatus{
			Healthy:   false,
			Connected: false,
			LatencyMs: time.Since(start).Milliseconds(),
			Error:     err.Error(),
		}, nil
	}
	return core.HealthStatus{
		Healthy:   exists,
		Connected: c.IsConnected(),
		LatencyMs: time.Since(start).Milliseconds(),
		Details: map[string]any{
			"subscription": c.cfg.Subscription.Name,
			"projectId":    c.cfg.Connection.ProjectID,
			"paused":       c.IsPaused(),
		},
	}, nil
}

var (
	_ core.Consumer      = (*Consumer)(nil)
	_ core.ConsumerHooks = (*Consumer)(nil)
)
